### ISSUE TITLE
Active Record CI for mysql2 shows "ActiveRecord::ConnectionNotEstablished: No connection pool"

### ISSUE THREAD

**Issue by yahonda** [rails/rails#41056]
Managed to reproduce CI failure at https://buildkite.com/rails/rails/builds/73935#824e38f8-a08d-4bcf-9195-302af0b481a6

### Steps to reproduce
```ruby
cd activerecord
ARCONN=mysql2 bin/test test/cases/connection_adapters/connection_swapping_nested_test.rb test/cases/sanitize_test.rb test/cases/multi_db_migrator_test.rb -n "/^(?:ActiveRecord::ConnectionAdapters::ConnectionSwappingNestedTest#(?:test_application_record_prevent_writes_can_be_changed)|SanitizeTest#(?:test_named_bind_arity)|MultiDbMigratorTest#(?:test_finds_migrations|test_finds_pending_migrations|test_get_all_versions|test_migrations_status|test_migrator_db_has_no_schema_migrations_table|test_migrator_forward|test_schema_migration_class_names))$/" --seed 22823
```

### Expected behavior
It should pass.

### Actual behavior
It always gets errors.

```ruby
$ ARCONN=mysql2 bin/test test/cases/connection_adapters/connection_swapping_nested_test.rb test/cases/sanitize_test.rb test/cases/multi_db_migrator_test.rb -n "/^(?:ActiveRecord::ConnectionAdapters::ConnectionSwappingNestedTest#(?:test_application_record_prevent_writes_can_be_changed)|SanitizeTest#(?:test_named_bind_arity)|MultiDbMigratorTest#(?:test_finds_migrations|test_finds_pending_migrations|test_get_all_versions|test_migrations_status|test_migrator_db_has_no_schema_migrations_table|test_migrator_forward|test_schema_migration_class_names))$/" --seed 22823
Using mysql2
Run options: -n "/^(?:ActiveRecord::ConnectionAdapters::ConnectionSwappingNestedTest#(?:test_application_record_prevent_writes_can_be_changed)|SanitizeTest#(?:test_named_bind_arity)|MultiDbMigratorTest#(?:test_finds_migrations|test_finds_pending_migrations|test_get_all_versions|test_migrations_status|test_migrator_db_has_no_schema_migrations_table|test_migrator_forward|test_schema_migration_class_names))$/" --seed 22823

# Running:

.E

Error:
MultiDbMigratorTest#test_migrator_forward:
ActiveRecord::ConnectionNotEstablished: No connection pool for 'ActiveRecord::ConnectionAdapters::ConnectionSwappingNestedTest::ApplicationRecord' found.
    /home/yahonda/src/github.com/rails/rails/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:1125:in `retrieve_connection'
    /home/yahonda/src/github.com/rails/rails/activerecord/lib/active_record/connection_handling.rb:303:in `retrieve_connection'
    /home/yahonda/src/github.com/rails/rails/activerecord/lib/active_record/connection_handling.rb:259:in `connection'
    /home/yahonda/src/github.com/rails/rails/activerecord/lib/active_record/schema_migration.rb:22:in `create_table'
    /home/yahonda/src/github.com/rails/rails/activerecord/test/cases/multi_db_migrator_test.rb:29:in `setup'


bin/test test/cases/multi_db_migrator_test.rb:170

E

Error:
MultiDbMigratorTest#test_finds_migrations:
ActiveRecord::ConnectionNotEstablished: No connection pool for 'ActiveRecord::ConnectionAdapters::ConnectionSwappingNestedTest::ApplicationRecord' found.
    /home/yahonda/src/github.com/rails/rails/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:1125:in `retrieve_connection'
    /home/yahonda/src/github.com/rails/rails/activerecord/lib/active_record/connection_handling.rb:303:in `retrieve_connection'
    /home/yahonda/src/github.com/rails/rails/activerecord/lib/active_record/connection_handling.rb:259:in `connection'
    /home/yahonda/src/github.com/rails/rails/activerecord/lib/active_record/schema_migration.rb:22:in `create_table'
    /home/yahonda/src/github.com/rails/rails/activerecord/test/cases/multi_db_migrator_test.rb:29:in `setup'


bin/test test/cases/multi_db_migrator_test.rb:75

E

Error:
MultiDbMigratorTest#test_schema_migration_class_names:
ActiveRecord::ConnectionNotEstablished: No connection pool for 'ActiveRecord::ConnectionAdapters::ConnectionSwappingNestedTest::ApplicationRecord' found.
    /home/yahonda/src/github.com/rails/rails/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:1125:in `retrieve_connection'
    /home/yahonda/src/github.com/rails/rails/activerecord/lib/active_record/connection_handling.rb:303:in `retrieve_connection'
    /home/yahonda/src/github.com/rails/rails/activerecord/lib/active_record/connection_handling.rb:259:in `connection'
    /home/yahonda/src/github.com/rails/rails/activerecord/lib/active_record/schema_migration.rb:22:in `create_table'
    /home/yahonda/src/github.com/rails/rails/activerecord/test/cases/multi_db_migrator_test.rb:29:in `setup'


bin/test test/cases/multi_db_migrator_test.rb:70

E

Error:
MultiDbMigratorTest#test_finds_pending_migrations:
ActiveRecord::ConnectionNotEstablished: No connection pool for 'ActiveRecord::ConnectionAdapters::ConnectionSwappingNestedTest::ApplicationRecord' found.
    /home/yahonda/src/github.com/rails/rails/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:1125:in `retrieve_connection'
    /home/yahonda/src/github.com/rails/rails/activerecord/lib/active_record/connection_handling.rb:303:in `retrieve_connection'
    /home/yahonda/src/github.com/rails/rails/activerecord/lib/active_record/connection_handling.rb:259:in `connection'
    /home/yahonda/src/github.com/rails/rails/activerecord/lib/active_record/schema_migration.rb:22:in `create_table'
    /home/yahonda/src/github.com/rails/rails/activerecord/test/cases/multi_db_migrator_test.rb:29:in `setup'


bin/test test/cases/multi_db_migrator_test.rb:136

E

Error:
MultiDbMigratorTest#test_get_all_versions:
ActiveRecord::ConnectionNotEstablished: No connection pool for 'ActiveRecord::ConnectionAdapters::ConnectionSwappingNestedTest::ApplicationRecord' found.
    /home/yahonda/src/github.com/rails/rails/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:1125:in `retrieve_connection'
    /home/yahonda/src/github.com/rails/rails/activerecord/lib/active_record/connection_handling.rb:303:in `retrieve_connection'
    /home/yahonda/src/github.com/rails/rails/activerecord/lib/active_record/connection_handling.rb:259:in `connection'
    /home/yahonda/src/github.com/rails/rails/activerecord/lib/active_record/schema_migration.rb:22:in `create_table'
    /home/yahonda/src/github.com/rails/rails/activerecord/test/cases/multi_db_migrator_test.rb:29:in `setup'


bin/test test/cases/multi_db_migrator_test.rb:107

E

Error:
MultiDbMigratorTest#test_migrations_status:
ActiveRecord::ConnectionNotEstablished: No connection pool for 'ActiveRecord::ConnectionAdapters::ConnectionSwappingNestedTest::ApplicationRecord' found.
    /home/yahonda/src/github.com/rails/rails/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:1125:in `retrieve_connection'
    /home/yahonda/src/github.com/rails/rails/activerecord/lib/active_record/connection_handling.rb:303:in `retrieve_connection'
    /home/yahonda/src/github.com/rails/rails/activerecord/lib/active_record/connection_handling.rb:259:in `connection'
    /home/yahonda/src/github.com/rails/rails/activerecord/lib/active_record/schema_migration.rb:22:in `create_table'
    /home/yahonda/src/github.com/rails/rails/activerecord/test/cases/multi_db_migrator_test.rb:29:in `setup'


bin/test test/cases/multi_db_migrator_test.rb:87

E

Error:
MultiDbMigratorTest#test_migrator_db_has_no_schema_migrations_table:
ActiveRecord::ConnectionNotEstablished: No connection pool for 'ActiveRecord::ConnectionAdapters::ConnectionSwappingNestedTest::ApplicationRecord' found.
    /home/yahonda/src/github.com/rails/rails/activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb:1125:in `retrieve_connection'
    /home/yahonda/src/github.com/rails/rails/activerecord/lib/active_record/connection_handling.rb:303:in `retrieve_connection'
    /home/yahonda/src/github.com/rails/rails/activerecord/lib/active_record/connection_handling.rb:259:in `connection'
    /home/yahonda/src/github.com/rails/rails/activerecord/lib/active_record/schema_migration.rb:22:in `create_table'
    /home/yahonda/src/github.com/rails/rails/activerecord/test/cases/multi_db_migrator_test.rb:29:in `setup'


bin/test test/cases/multi_db_migrator_test.rb:152

.

Finished in 0.037471s, 240.1869 runs/s, 80.0623 assertions/s.
9 runs, 3 assertions, 0 failures, 7 errors, 0 skips
$
```

### System configuration
**Rails version**:master branch

**Ruby version**:ruby 3.0.0p0 (2020-12-25 revision 95aff21468) [x86_64-linux]
It looks like these errors are not related to the Ruby version.

**Comment by yahonda** [comment]
It also reproduces with other bundled `postgresql` and `sqlite3` adapters.

**Comment by yahonda** [comment]
I've found these errors are triggered by #41030.

**Comment by eileencodes** [comment]
Oops I swear the build was marked as green. Working on a fix.

**Comment by eileencodes** [comment]
Should be fixed by 67fd5cb3efe6cbd31ad88261abe9a03f186811c5 on 6-1-stable and eeb3879a114769f7b2c8470603cc06c61ed3c5f0 on master. Thanks for the easy repro @yahonda!
